pyriemann/utils/mean.py

```diff
@@ -733,3 +734,144 @@ def nanmean_riemann(covmats, tol=10e-9, maxiter=100, init=None,
+
+    Parameters
+    ----------
+    covmats : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of matrices.
+    tol : float, default=10e-6
+        The tolerance to stop the gradient descent.
+    maxiter : int, default=50
+        The maximum number of iterations.
+    init : None | ndarray, shape (n_channels, n_channels), default=None
+        A matrix used to initialize the gradient descent.
+        If None, the Euclidean mean is used.
+
+    Returns
+    -------
+    C : ndarray, shape (n_channels, n_channels)
+        Euclidean geometric median.
+
+    Notes
+    -----
+    .. versionadded:: 0.3.1
+
+    References
+    ----------
+    .. [1] Vardi Y and Zhan C-H. "The multivariate L1-median and associated
+        data depth", PNAS, 2000
+    """
+    if init is None:
+        C = mean_euclid(covmats)
+    else:
+        C = init
+
+    for _ in range(maxiter):
+        dists = pairwise_distance(
+            covmats,
+            C[np.newaxis, ...],
+            metric='euclid'
+        )[:, 0]
+        is_zero = (dists == 0)
+
+        inv_dists = 1 / dists[~is_zero]
+        weights = inv_dists / np.sum(inv_dists)
+        Cnew = np.einsum('a,abc->bc', weights, covmats[~is_zero])
+
+        n_zeros = np.sum(is_zero)
+        if n_zeros > 0:
+            r = np.linalg.norm((Cnew - C) * np.sum(inv_dists), ord='fro')
+            rinv = 0 if r == 0 else n_zeros / r
+            Cnew = max(0, 1 - rinv) * Cnew + min(1, rinv) * C
+
+        crit = np.linalg.norm(Cnew - C, ord='fro')
+        C = Cnew
+        if crit <= tol:
+            break
+    else:
+        warnings.warn('Convergence not reached')
+
+    return C
+
+
+def median_riemann(covmats, tol=10e-6, maxiter=50, init=None):
+    """Median of SPD matrices according to the Riemannian metric.
+
+    Return the affine-invariant Riemannian geometric median for SPD matrices.
+
+    Parameters
+    ----------
+    covmats : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of SPD matrices.
+    tol : float, default=10e-6
+        The tolerance to stop the gradient descent.
+    maxiter : int, default=50
+        The maximum number of iterations.
+    init : None | ndarray, shape (n_channels, n_channels), default=None
+        A SPD matrix used to initialize the gradient descent.
+        If None, the Euclidean mean is used.
+
+    Returns
+    -------
+    C : ndarray, shape (n_channels, n_channels)
+        Affine-invariant Riemannian geometric median.
+
+    Notes
+    -----
+    .. versionadded:: 0.3.1
+
+    References
+    ----------
+    .. [1] Fletcher PT, Venkatasubramanian S and Joshi S. "The geometric median
+        on Riemannian manifolds with application to robust atlas estimation",
+        NeuroImage, 2009
+    .. [2] Yang L, Arnaudon M and Barbaresco F. "Riemannian median, geometry of
+        covariance matrices and radar target detection", EURAD, 2010
+    """
+    if init is None:
+        C = mean_euclid(covmats)
```

Comment:
why not mean_riemann?

Reply:
I don’t know if it’s a safe idea to initialize an iterative algorithm using another iterative algorithm that may not converge.